saw-core/src/Verifier/SAW/SharedTerm.hs

```diff
@@ -339,6 +340,7 @@ data SharedContext = SharedContext
   { scModuleMap      :: IORef ModuleMap
   , scTermF          :: TermF Term -> IO Term
   , scNamingEnv      :: IORef SAWNamingEnv
```

Comment:
Should we be using `MVar` everywhere instead of `IORef`? If not, what's the reasoning behind choosing one over the other?

Reply:
I was actually wondering the same thing myself; I don't know any good guidelines for when to choose `IORef` vs `MVar`. I only chose `MVar` because I thought I needed to do some `IO` in the middle of my updating the environment (so that I could throw an exception in `IO` if there is a duplicate entry) and I didn't seem to be able to do that with `modifyIORef`. But I see now that `atomicModifyIORef` has a different type that would have worked just fine for my purpose.

Reply:
I guess I'd prefer not to mix, unless we have a good reason.  So far, I don't know that we have a very well-reasoned multithreaded data coherence story, but `IORef` plus atomic modifications at least prevents the most egregious kinds data races.  I'm aways wary of bare `MVar`s because it is so easy to deadlock.

Reply:
I'll switch it to `IORef` with `atomicModifyIORef'`.